pkg/logging/rate_limited_logger.go

```diff
@@ -1,128 +1,288 @@
 func (rl *RateLimitedLogger) Debugf(template string, args ...interface{}) {
-	rl.Logger.Debugf(template, args...)
+	rl.logger.Debugf(template, args...)
 }
 
 // Debugw logs the input message and keyValues
 func (rl *RateLimitedLogger) Debugw(msg string, keysAndValues ...interface{}) {
-	rl.Logger.Debugw(msg, keysAndValues...)
+	rl.logger.Debugw(msg, keysAndValues...)
+}
+
+func getLogKey(limiter string, level zapcore.Level, file string, line int, payload string) string {
+	var keyWriter strings.Builder
+	keyWriter.WriteString(limiter)
+	keyWriter.WriteString("-")
+	keyWriter.WriteString(level.CapitalString())
+	keyWriter.WriteString("-")
+	keyWriter.WriteString(file)
+	keyWriter.WriteString(":")
+	keyWriter.WriteString(fmt.Sprintf("%d", line))
+	keyWriter.WriteString("-")
+	keyWriter.WriteString(payload)
+	return keyWriter.String()
+}
+
+const (
+	localFilePathPrefix = "github.com/stackrox/stackrox/"
+	filePathPrefix      = "github.com/stackrox/rox/"
+	githubPathPrefix    = "/__w/stackrox/stackrox/"
+)
+
+func getTrimmedFilePath(path string) string {
+	prefixes := []string{filePathPrefix, localFilePathPrefix, githubPathPrefix}
+	cutPath := path
+	for _, prefix := range prefixes {
+		prefixToCut := strings.Index(cutPath, prefix)
+		if prefixToCut >= 0 {
+			cutPath = cutPath[prefixToCut:]
+		}
+	}
+	trimmedPath := cutPath
+	for _, prefix := range prefixes {
+		trimmedPath = strings.TrimPrefix(trimmedPath, prefix)
+	}
+	return trimmedPath
+}
+
+func (rl *RateLimitedLogger) registerTraceAndLog(
+	level zapcore.Level,
+	limiter string,
+	key string,
+	payload string,
+	file string,
+	line int,
+) {
+	log := newRateLimitedLog(
+		rl.logger,
+		level,
+		limiter,
+		payload,
+		file,
+		line,
+	)
+	rl.rateLimitedLogs.Add(key, log)
+	log.log()
+}
+
+func (rl *RateLimitedLogger) logf(level zapcore.Level, limiter string, template string, args ...interface{}) {
+	payload := fmt.Sprintf(template, args...)
+	_, file, line, ok := runtime.Caller(2)
+	if !ok {
+		file, line = "", 0
+	}
+	file = getTrimmedFilePath(file)
+	key := getLogKey(limiter, level, file, line, payload)
+	if throttledLog, found := rl.rateLimitedLogs.Get(key); found && throttledLog != nil {
+		throttledLog.count.Add(1)
+	} else if found && throttledLog == nil {
+		// There is something wrong in the cache. Clean up.
+		rl.rateLimitedLogs.Remove(key)
+		rl.registerTraceAndLog(level, limiter, key, payload, file, line)
+	} else {
+		rl.registerTraceAndLog(level, limiter, key, payload, file, line)
+	}
+}
+
+func (rl *RateLimitedLogger) stop() {
+	// Flush logs
+	rl.rateLimitedLogs.Purge()
+	// Stop background thread
+	rl.rateLimitedLogs.Close()
+}
+
+func stopLogger(logger *RateLimitedLogger) {
+	logger.stop()
+}
+
+const (
+	limitedLogSuffixFormat = " - %d log suppressed for limiter %q"
+)
+
+type rateLimitedLog struct {
+	logger   Logger
+	level    zapcore.Level
+	limiter  string
+	payload  string
+	file     string
+	line     int
+	count    atomic.Int32
+	logMutex sync.Mutex
```

Comment:
I have noticed that this mutex is used only in the `log` function. I don't see a reason for using it. The only situation where that would be useful is when a log is created and evicted after that, but in that case, eviction would not call a log because the count is 0.

@rhybrillou Can you explain to me why need this lock? It's really difficult for me to understand race conditions where this mutex would help us.

I have more concerns about the section in `logf` where we can increase the count for a log that is potentially evicted:
```
	if throttledLog, found := rl.rateLimitedLogs.Get(key); found && throttledLog != nil {
		throttledLog.count.Add(1)
```

Reply:
The mutex was added some time ago. I think back then, the problem was race conditions reported by unit tests when expiration was added as a background go routine.
It is removed now. The log function acts based on log count, and the atomic int for count is the main gate to protect against race conditions.
I thought further about the case you mention, and added an `id` to cached rate-limited log, along with a post-update check telling whether the updated item is still in cache. I think with that most potential log loss scenarios are covered. The last one to cover would be the cache flush on shutdown case.